TLS Protocols & Ciphers with Spring Boot

This is related to issue #1659 

HTTPS is working properly until I set armeria.ssl.enabled-protocols or armeria.ssl.ciphers

Here is my spring boot configuration
```
server:
  port: 9411
  ssl:
    enabled: true
    key-alias: "myhost.mycompany.fr"
    # ciphers: "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,TLS_DHE_RSA_WITH_AES_256_CBC_SHA256,TLS_DHE_RSA_WITH_AES_128_CBC_SHA256,TLS_ECDH_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA384,TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256,TLS_ECDH_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384,TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_256_CBC_SHA256,TLS_RSA_WITH_AES_128_CBC_SHA256"
    #enabled-protocols: "TLSv1.2"
    key-store: "/opt/zipkin-2.12.8/keystore.jks"
    key-store-password: "changeme"
    trust-store: "/opt/zipkin-2.12.8/keystore.jks"
    trust-store-password: "changeme"

armeria:
  ports:
    - port: 9411
      protocol: HTTPS
  ssl:
    key-alias: "myhost.mycompany.fr"
    # ciphers: "TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,TLS_DHE_RSA_WITH_AES_256_CBC_SHA256,TLS_DHE_RSA_WITH_AES_128_CBC_SHA256,TLS_ECDH_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA384,TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256,TLS_ECDH_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384,TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_256_CBC_SHA256,TLS_RSA_WITH_AES_128_CBC_SHA256"
    #enabled-protocols: "TLSv1.2"
    key-store: "/opt/zipkin-2.12.8/keystore.jks"
    key-store-password: "changeme"
    trust-store: "/opt/zipkin-2.12.8/keystore.jks"
    trust-store-password: "changeme"
```

When I uncomment enabled-protocols:
```
2019-04-11T18:01:29,975 WARN  com.linecorp.armeria.server.HttpServerPipelineConfigurator$Http2OrHttpHandler - [id: 0xd9cf80d2, L:/10.122.135.222:9411 - R:/163.113.65.61:48292] TLS handshake failed:
javax.net.ssl.SSLHandshakeException: error:1000009c:SSL routines:OPENSSL_internal:HTTP_REQUEST
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.shutdownWithError(ReferenceCountedOpenSslEngine.java:973) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.sslReadErrorResult(ReferenceCountedOpenSslEngine.java:1231) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1185) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1256) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1299) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.SslHandler$SslEngineType$1.unwrap(SslHandler.java:217) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1330) [netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.SslHandler.decodeNonJdkCompatible(SslHandler.java:1237) [netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1274) [netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502) [netty-codec-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441) [netty-codec-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278) [netty-codec-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.handlerRemoved(ByteToMessageDecoder.java:249) [netty-codec-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507) [netty-codec-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441) [netty-codec-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278) [netty-codec-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.flush.FlushConsolidationHandler.channelRead(FlushConsolidationHandler.java:154) [netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:796) [netty-transport-native-epoll-4.1.34.Final-linux-x86_64.jar!/:4.1.34.Final]
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:427) [netty-transport-native-epoll-4.1.34.Final-linux-x86_64.jar!/:4.1.34.Final]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:328) [netty-transport-native-epoll-4.1.34.Final-linux-x86_64.jar!/:4.1.34.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) [netty-common-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.34.Final.jar!/:4.1.34.Final]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_202]
```

When I uncomment ciphers (but not protocols):
```
Caused by: java.lang.IllegalArgumentException: unsupported cipher suite: TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384(ECDHE-ECDSA-AES256-SHA384)
        at io.netty.handler.ssl.CipherSuiteConverter.convertToCipherStrings(CipherSuiteConverter.java:470) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.ReferenceCountedOpenSslContext.<init>(ReferenceCountedOpenSslContext.java:254) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.OpenSslContext.<init>(OpenSslContext.java:43) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.OpenSslServerContext.<init>(OpenSslServerContext.java:345) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.OpenSslServerContext.<init>(OpenSslServerContext.java:333) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.SslContext.newServerContextInternal(SslContext.java:423) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.SslContextBuilder.build(SslContextBuilder.java:447) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at com.linecorp.armeria.internal.spring.ArmeriaConfigurationUtil.configureTls(ArmeriaConfigurationUtil.java:383) ~[armeria-spring-boot-autoconfigure-0.83.0.jar!/:?]
        at com.linecorp.armeria.internal.spring.ArmeriaConfigurationUtil.configureTls(ArmeriaConfigurationUtil.java:343) ~[armeria-spring-boot-autoconfigure-0.83.0.jar!/:?]
        at com.linecorp.armeria.internal.spring.ArmeriaConfigurationUtil.configureServerWithArmeriaSettings(ArmeriaConfigurationUtil.java:163) ~[armeria-spring-boot-autoconfigure-0.83.0.jar!/:?]
        at com.linecorp.armeria.spring.ArmeriaAutoConfiguration.armeriaServer(ArmeriaAutoConfiguration.java:105) ~[armeria-spring-boot-autoconfigure-0.83.0.jar!/:?]
        at com.linecorp.armeria.spring.ArmeriaAutoConfiguration$$EnhancerBySpringCGLIB$$16ba9c43.CGLIB$armeriaServer$0(<generated>) ~[armeria-spring-boot-autoconfigure-0.83.0.jar!/:?]
        at com.linecorp.armeria.spring.ArmeriaAutoConfiguration$$EnhancerBySpringCGLIB$$16ba9c43$$FastClassBySpringCGLIB$$ad64608f.invoke(<generated>) ~[armeria-spring-boot-autoconfigure-0.83.0.jar!/:?]
        at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244) ~[spring-core-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
        at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:363) ~[spring-context-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
        at com.linecorp.armeria.spring.ArmeriaAutoConfiguration$$EnhancerBySpringCGLIB$$16ba9c43.armeriaServer(<generated>) ~[armeria-spring-boot-autoconfigure-0.83.0.jar!/:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_202]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_202]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_202]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_202]
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
        at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:622) ~[spring-beans-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
        ... 25 more
```

What are the appropriate values for fields ciphers and enabled-protocols ?
* SSL Protocols https://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html#SSLContext
* SSL Cipher Suites https://docs.oracle.com/javase/8/docs/technotes/guides/security/SunProviders.html#SupportedCipherSuites

We are running Java 8 u202 which supports cipher TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384
```
$  java  -version
openjdk version "1.8.0_202"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_202-b08)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.202-b08, mixed mode)
```

```
2019-04-12T14:02:01,241 DEBUG org.springframework.boot.StartupInfoLogger - Running with Spring Boot v2.1.4.RELEASE, Spring v5.1.6.RELEASE
2019-04-12T14:02:01,242 INFO  org.springframework.boot.SpringApplication - The following profiles are active: shared
2019-04-12T14:02:04,124 INFO  com.linecorp.armeria.common.Flags - com.linecorp.armeria.verboseExceptions: false (default)
2019-04-12T14:02:04,125 INFO  com.linecorp.armeria.common.Flags - com.linecorp.armeria.verboseSocketExceptions: false (default)
2019-04-12T14:02:04,126 INFO  com.linecorp.armeria.common.Flags - com.linecorp.armeria.verboseResponses: false (default)
2019-04-12T14:02:04,160 INFO  com.linecorp.armeria.common.Flags - com.linecorp.armeria.useEpoll: true (default)
2019-04-12T14:02:04,296 INFO  com.linecorp.armeria.common.Flags - com.linecorp.armeria.useOpenSsl: false
2019-04-12T14:02:04,297 INFO  com.linecorp.armeria.common.Flags - com.linecorp.armeria.maxNumConnections: 2147483647 (default)
2019-04-12T14:02:04,297 INFO  com.linecorp.armeria.common.Flags - com.linecorp.armeria.numCommonWorkers: 8 (default)
2019-04-12T14:02:04,298 INFO  com.linecorp.armeria.common.Flags - com.linecorp.armeria.numCommonBlockingTaskThreads: 200 (default)
```
The error is the same:
```
Caused by: java.lang.IllegalArgumentException: unsupported cipher suite: TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384(ECDHE-ECDSA-AES256-SHA384)
        at io.netty.handler.ssl.CipherSuiteConverter.convertToCipherStrings(CipherSuiteConverter.java:470) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.ReferenceCountedOpenSslContext.<init>(ReferenceCountedOpenSslContext.java:254) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.OpenSslContext.<init>(OpenSslContext.java:43) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.OpenSslServerContext.<init>(OpenSslServerContext.java:345) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.OpenSslServerContext.<init>(OpenSslServerContext.java:333) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.SslContext.newServerContextInternal(SslContext.java:423) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at io.netty.handler.ssl.SslContextBuilder.build(SslContextBuilder.java:447) ~[netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
        at com.linecorp.armeria.internal.spring.ArmeriaConfigurationUtil.configureTls(ArmeriaConfigurationUtil.java:383) ~[armeria-spring-boot-autoconfigure-0.83.0.jar!/:?]
        at com.linecorp.armeria.internal.spring.ArmeriaConfigurationUtil.configureTls(ArmeriaConfigurationUtil.java:343) ~[armeria-spring-boot-autoconfigure-0.83.0.jar!/:?]
        at com.linecorp.armeria.internal.spring.ArmeriaConfigurationUtil.configureServerWithArmeriaSettings(ArmeriaConfigurationUtil.java:163) ~[armeria-spring-boot-autoconfigure-0.83.0.jar!/:?]
```

--- a/core/src/main/java/com/linecorp/armeria/server/AbstractVirtualHostBuilder.java
+++ b/core/src/main/java/com/linecorp/armeria/server/AbstractVirtualHostBuilder.java
@@ -27,9 +27,12 @@ import java.nio.charset.Charset;
 import java.security.cert.CertificateException;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.function.Consumer;
 import java.util.function.Function;
+import java.util.function.Supplier;
 
 import javax.annotation.Nullable;
+import javax.net.ssl.KeyManagerFactory;
 import javax.net.ssl.SSLException;
 
 import org.slf4j.Logger;
@@ -153,7 +156,17 @@ abstract class AbstractVirtualHostBuilder<B extends AbstractVirtualHostBuilder> 
      * and cleartext {@code keyFile}.
      */
     public B tls(File keyCertChainFile, File keyFile) throws SSLException {
-        tls(keyCertChainFile, keyFile, null);
+        tls(keyCertChainFile, keyFile, null, sslContextBuilder -> { /* noop */ });
+        return self();
+    }
+
+    /**
+     * Configures SSL or TLS of this {@link VirtualHost} with the specified {@code keyCertChainFile},
+     * cleartext {@code keyFile} and {@code tlsCustomizer}.
+     */
+    public B tls(File keyCertChainFile, File keyFile, Consumer<SslContextBuilder> tlsCustomizer)
+            throws SSLException {
+        tls(keyCertChainFile, keyFile, null, tlsCustomizer);
         return self();
     }
 
@@ -162,6 +175,20 @@ abstract class AbstractVirtualHostBuilder<B extends AbstractVirtualHostBuilder> 
      * {@code keyFile} and {@code keyPassword}.
      */
     public B tls(File keyCertChainFile, File keyFile, @Nullable String keyPassword) throws SSLException {
+        tls(keyCertChainFile, keyFile, keyPassword, sslContextBuilder -> { /* noop */ });
+        return self();
+    }
+
+    /**
+     * Configures SSL or TLS of this {@link VirtualHost} with the specified {@code keyCertChainFile},
+     * {@code keyFile}, {@code keyPassword} and {@code tlsCustomizer}.
+     */
+    public B tls(File keyCertChainFile, File keyFile, @Nullable String keyPassword,
+                 Consumer<SslContextBuilder> tlsCustomizer) throws SSLException {
+        requireNonNull(keyCertChainFile, "keyCertChainFile");
+        requireNonNull(keyFile, "keyFile");
+        requireNonNull(tlsCustomizer, "tlsCustomizer");
+
         if (!keyCertChainFile.exists()) {
             throw new SSLException("non-existent certificate chain file: " + keyCertChainFile);
         }
@@ -175,26 +202,21 @@ abstract class AbstractVirtualHostBuilder<B extends AbstractVirtualHostBuilder> 
             throw new SSLException("cannot read key file: " + keyFile);
         }
 
-        final SslContext sslCtx;
+        tls(buildSslContext(() -> SslContextBuilder.forServer(keyCertChainFile, keyFile, keyPassword),
+                            tlsCustomizer));
+        return self();
+    }
 
-        try {
-            sslCtx = BouncyCastleKeyFactoryProvider.call(() -> {
-                final SslContextBuilder builder =
-                        SslContextBuilder.forServer(keyCertChainFile, keyFile, keyPassword);
+    /**
+     * Configures SSL or TLS of this {@link VirtualHost} with the specified {@code keyManagerFactory}
+     * and {@code tlsCustomizer}.
+     */
+    public B tls(KeyManagerFactory keyManagerFactory,
+                 Consumer<SslContextBuilder> tlsCustomizer) throws SSLException {
+        requireNonNull(keyManagerFactory, "keyManagerFactory");
+        requireNonNull(tlsCustomizer, "tlsCustomizer");
 
-                builder.sslProvider(Flags.useOpenSsl() ? SslProvider.OPENSSL : SslProvider.JDK);
-                builder.ciphers(Http2SecurityUtil.CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
-                builder.applicationProtocolConfig(HTTPS_ALPN_CFG);
-
-                return builder.build();
-            });
-        } catch (RuntimeException | SSLException e) {
-            throw e;
-        } catch (Exception e) {
-            throw new SSLException("failed to configure TLS: " + e, e);
-        }
-
-        tls(sslCtx);
+        tls(buildSslContext(() -> SslContextBuilder.forServer(keyManagerFactory), tlsCustomizer));
         return self();
     }
 
@@ -207,6 +229,28 @@ abstract class AbstractVirtualHostBuilder<B extends AbstractVirtualHostBuilder> 
     public B tlsSelfSigned() throws SSLException, CertificateException {
         final SelfSignedCertificate ssc = new SelfSignedCertificate(defaultHostname);
         return tls(ssc.certificate(), ssc.privateKey());
+    }
+
+    private static SslContext buildSslContext(
+            Supplier<SslContextBuilder> builderSupplier,
+            Consumer<SslContextBuilder> tlsCustomizer) throws SSLException {
+        try {
+            return BouncyCastleKeyFactoryProvider.call(() -> {
+                final SslContextBuilder builder = builderSupplier.get();
+
+                builder.sslProvider(Flags.useOpenSsl() ? SslProvider.OPENSSL : SslProvider.JDK);
+                builder.ciphers(Http2SecurityUtil.CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
+                builder.applicationProtocolConfig(HTTPS_ALPN_CFG);
+
+                tlsCustomizer.accept(builder);
+
+                return builder.build();
+            });
+        } catch (RuntimeException | SSLException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new SSLException("failed to configure TLS: " + e, e);
+        }
     }
 
     /**
@@ -316,14 +360,14 @@ abstract class AbstractVirtualHostBuilder<B extends AbstractVirtualHostBuilder> 
      */
     public B service(ServiceWithPathMappings<HttpRequest, HttpResponse> serviceWithPathMappings,
                      Iterable<Function<? super Service<HttpRequest, HttpResponse>,
-                                       ? extends Service<HttpRequest, HttpResponse>>> decorators) {
+                             ? extends Service<HttpRequest, HttpResponse>>> decorators) {
         requireNonNull(serviceWithPathMappings, "serviceWithPathMappings");
         requireNonNull(serviceWithPathMappings.pathMappings(), "serviceWithPathMappings.pathMappings()");
         requireNonNull(decorators, "decorators");
 
         Service<HttpRequest, HttpResponse> decorated = serviceWithPathMappings;
         for (Function<? super Service<HttpRequest, HttpResponse>,
-                      ? extends Service<HttpRequest, HttpResponse>> d : decorators) {
+                ? extends Service<HttpRequest, HttpResponse>> d : decorators) {
             checkNotNull(d, "decorators contains null: %s", decorators);
             decorated = d.apply(decorated);
             checkNotNull(decorated, "A decorator returned null: %s", d);
@@ -344,7 +388,7 @@ abstract class AbstractVirtualHostBuilder<B extends AbstractVirtualHostBuilder> 
     @SafeVarargs
     public final B service(ServiceWithPathMappings<HttpRequest, HttpResponse> serviceWithPathMappings,
                            Function<? super Service<HttpRequest, HttpResponse>,
-                                    ? extends Service<HttpRequest, HttpResponse>>... decorators) {
+                                   ? extends Service<HttpRequest, HttpResponse>>... decorators) {
         return service(serviceWithPathMappings, ImmutableList.copyOf(requireNonNull(decorators, "decorators")));
     }
 

--- a/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java
+++ b/core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java
@@ -41,11 +41,13 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.Executor;
+import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
 import javax.annotation.Nullable;
+import javax.net.ssl.KeyManagerFactory;
 import javax.net.ssl.SSLException;
 
 import org.slf4j.Logger;
@@ -76,6 +78,7 @@ import io.netty.channel.ChannelOption;
 import io.netty.channel.EventLoopGroup;
 import io.netty.channel.epoll.EpollChannelOption;
 import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
 import io.netty.util.DomainNameMapping;
 import io.netty.util.DomainNameMappingBuilder;
 import io.netty.util.concurrent.GlobalEventExecutor;
@@ -731,6 +734,20 @@ public final class ServerBuilder {
 
     /**
      * Configures SSL or TLS of the default {@link VirtualHost} from the specified {@code keyCertChainFile},
+     * cleartext {@code keyFile} and {@code tlsCustomizer}.
+     *
+     * @throws IllegalStateException if the default {@link VirtualHost} has been set via
+     *                               {@link #defaultVirtualHost(VirtualHost)} already
+     */
+    public ServerBuilder tls(File keyCertChainFile, File keyFile,
+                             Consumer<SslContextBuilder> tlsCustomizer) throws SSLException {
+        defaultVirtualHostBuilderUpdated();
+        defaultVirtualHostBuilder.tls(keyCertChainFile, keyFile, tlsCustomizer);
+        return this;
+    }
+
+    /**
+     * Configures SSL or TLS of the default {@link VirtualHost} from the specified {@code keyCertChainFile},
      * {@code keyFile} and {@code keyPassword}.
      *
      * @throws IllegalStateException if the default {@link VirtualHost} has been set via
@@ -738,9 +755,38 @@ public final class ServerBuilder {
      */
     public ServerBuilder tls(
             File keyCertChainFile, File keyFile, @Nullable String keyPassword) throws SSLException {
-
         defaultVirtualHostBuilderUpdated();
         defaultVirtualHostBuilder.tls(keyCertChainFile, keyFile, keyPassword);
+        return this;
+    }
+
+    /**
+     * Configures SSL or TLS of the default {@link VirtualHost} from the specified {@code keyCertChainFile},
+     * {@code keyFile}, {@code keyPassword} and {@code tlsCustomizer}.
+     *
+     * @throws IllegalStateException if the default {@link VirtualHost} has been set via
+     *                               {@link #defaultVirtualHost(VirtualHost)} already
+     */
+    public ServerBuilder tls(
+            File keyCertChainFile, File keyFile, @Nullable String keyPassword,
+            Consumer<SslContextBuilder> tlsCustomizer) throws SSLException {
+        defaultVirtualHostBuilderUpdated();
+        defaultVirtualHostBuilder.tls(keyCertChainFile, keyFile, keyPassword, tlsCustomizer);
+        return this;
+    }
+
+    /**
+     * Configures SSL or TLS of the default {@link VirtualHost} from the specified {@code keyManagerFactory}
+     * and {@code tlsCustomizer}.
+     *
+     * @throws IllegalStateException if the default {@link VirtualHost} has been set via
+     *                               {@link #defaultVirtualHost(VirtualHost)} already
+     */
+    public ServerBuilder tls(KeyManagerFactory keyManagerFactory,
+                             Consumer<SslContextBuilder> tlsCustomizer) throws SSLException {
+
+        defaultVirtualHostBuilderUpdated();
+        defaultVirtualHostBuilder.tls(keyManagerFactory, tlsCustomizer);
         return this;
     }
 
@@ -895,7 +941,7 @@ public final class ServerBuilder {
     public ServerBuilder service(
             ServiceWithPathMappings<HttpRequest, HttpResponse> serviceWithPathMappings,
             Iterable<Function<? super Service<HttpRequest, HttpResponse>,
-                              ? extends Service<HttpRequest, HttpResponse>>> decorators) {
+                    ? extends Service<HttpRequest, HttpResponse>>> decorators) {
         defaultVirtualHostBuilderUpdated();
         defaultVirtualHostBuilder.service(serviceWithPathMappings, decorators);
         return this;
@@ -914,7 +960,7 @@ public final class ServerBuilder {
     public final ServerBuilder service(
             ServiceWithPathMappings<HttpRequest, HttpResponse> serviceWithPathMappings,
             Function<? super Service<HttpRequest, HttpResponse>,
-                     ? extends Service<HttpRequest, HttpResponse>>... decorators) {
+                    ? extends Service<HttpRequest, HttpResponse>>... decorators) {
         defaultVirtualHostBuilderUpdated();
         defaultVirtualHostBuilder.service(serviceWithPathMappings, decorators);
         return this;

--- a/spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/internal/spring/ArmeriaConfigurationUtil.java
+++ b/spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/internal/spring/ArmeriaConfigurationUtil.java
@@ -95,6 +95,7 @@ import io.micrometer.core.instrument.composite.CompositeMeterRegistry;
 import io.micrometer.core.instrument.dropwizard.DropwizardMeterRegistry;
 import io.micrometer.prometheus.PrometheusMeterRegistry;
 import io.netty.handler.ssl.ClientAuth;
+import io.netty.handler.ssl.SslProvider;
 import io.netty.handler.ssl.SupportedCipherSuiteFilter;
 import io.netty.util.NetUtil;
 import io.prometheus.client.CollectorRegistry;
@@ -389,6 +390,10 @@ public final class ArmeriaConfigurationUtil {
             sb.tls(keyManagerFactory, sslContextBuilder -> {
                 sslContextBuilder.trustManager(trustManagerFactory);
 
+                final SslProvider sslProvider = ssl.getProvider();
+                if (sslProvider != null) {
+                    sslContextBuilder.sslProvider(sslProvider);
+                }
                 final List<String> enabledProtocols = ssl.getEnabledProtocols();
                 if (enabledProtocols != null) {
                     sslContextBuilder.protocols(enabledProtocols.toArray(new String[enabledProtocols.size()]));

--- a/spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/internal/spring/ArmeriaConfigurationUtil.java
+++ b/spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/internal/spring/ArmeriaConfigurationUtil.java
@@ -63,7 +63,6 @@ import com.google.common.base.Strings;
 import com.google.common.collect.ImmutableList;
 import com.google.common.math.LongMath;
 
-import com.linecorp.armeria.common.Flags;
 import com.linecorp.armeria.common.HttpHeaderNames;
 import com.linecorp.armeria.common.HttpHeaders;
 import com.linecorp.armeria.common.HttpRequest;
@@ -96,8 +95,7 @@ import io.micrometer.core.instrument.composite.CompositeMeterRegistry;
 import io.micrometer.core.instrument.dropwizard.DropwizardMeterRegistry;
 import io.micrometer.prometheus.PrometheusMeterRegistry;
 import io.netty.handler.ssl.ClientAuth;
-import io.netty.handler.ssl.SslContextBuilder;
-import io.netty.handler.ssl.SslProvider;
+import io.netty.handler.ssl.SupportedCipherSuiteFilter;
 import io.netty.util.NetUtil;
 import io.prometheus.client.CollectorRegistry;
 
@@ -385,31 +383,26 @@ public final class ArmeriaConfigurationUtil {
                 return;
             }
 
-            final SslContextBuilder sslBuilder = SslContextBuilder
-                    .forServer(getKeyManagerFactory(ssl, keyStoreSupplier))
-                    .trustManager(getTrustManagerFactory(ssl, trustStoreSupplier));
-            final SslProvider sslProvider = ssl.getProvider() == null ?
-                                            Flags.useOpenSsl() ?
-                                            SslProvider.OPENSSL
-                                                               : SslProvider.JDK
-                                                                      : ssl.getProvider();
-            sslBuilder.sslProvider(sslProvider);
-            final List<String> enabledProtocols = ssl.getEnabledProtocols();
-            if (enabledProtocols != null) {
-                sslBuilder.protocols(enabledProtocols.toArray(new String[enabledProtocols.size()]));
-            }
+            final KeyManagerFactory keyManagerFactory = getKeyManagerFactory(ssl, keyStoreSupplier);
+            final TrustManagerFactory trustManagerFactory = getTrustManagerFactory(ssl, trustStoreSupplier);
 
-            final List<String> ciphers = ssl.getCiphers();
-            if (ciphers != null) {
-                sslBuilder.ciphers(ImmutableList.copyOf(ciphers));
-            }
+            sb.tls(keyManagerFactory, sslContextBuilder -> {
+                sslContextBuilder.trustManager(trustManagerFactory);
 
-            final ClientAuth clientAuth = ssl.getClientAuth();
-            if (clientAuth != null) {
-                sslBuilder.clientAuth(clientAuth);
-            }
-
-            sb.tls(sslBuilder.build());
+                final List<String> enabledProtocols = ssl.getEnabledProtocols();
+                if (enabledProtocols != null) {
+                    sslContextBuilder.protocols(enabledProtocols.toArray(new String[enabledProtocols.size()]));
+                }
+                final List<String> ciphers = ssl.getCiphers();
+                if (ciphers != null) {
+                    sslContextBuilder.ciphers(ImmutableList.copyOf(ciphers),
+                                              SupportedCipherSuiteFilter.INSTANCE);
+                }
+                final ClientAuth clientAuth = ssl.getClientAuth();
+                if (clientAuth != null) {
+                    sslContextBuilder.clientAuth(clientAuth);
+                }
+            });
         } catch (Exception e) {
             throw new IllegalStateException("Failed to configure TLS: " + e, e);
         }

